src/Chainweb/Version.hs

```diff
@@ -483,61 +459,138 @@ instance HasChainwebVersion ChainwebVersion where
     _chainwebVersion = id
     {-# INLINE _chainwebVersion #-}
 
+-- | All known chainIds. This includes chains that are not yet "active".
+--
 chainIds :: HasChainwebVersion v => v -> HS.HashSet ChainId
-chainIds = graphChainIds . _chainGraph . _chainwebVersion
+chainIds = graphChainIds . snd . NE.head . chainwebGraphs . _chainwebVersion
 {-# INLINE chainIds #-}
 
 mkChainId
     :: MonadThrow m
     => HasChainwebVersion v
     => Integral i
     => v
+    -> BlockHeight
     -> i
     -> m ChainId
-mkChainId v i = cid
-    <$ checkWebChainId (chainwebVersionGraph $ _chainwebVersion v) cid
+mkChainId v h i = cid
+    <$ checkWebChainId (chainGraphAt (_chainwebVersion v) h) cid
   where
     cid = unsafeChainId (fromIntegral i)
 {-# INLINE mkChainId #-}
 
--- | Sometimes, in particular for testing and examples, some fixed chain id is
--- needed, but it doesn't matter which one. This function provides some valid
--- chain ids.
---
-someChainId :: HasCallStack => HasChainwebVersion v => v -> ChainId
-someChainId = head . toList . chainIds
-    -- 'head' is guaranteed to succeed because the empty graph isn't a valid chain
-    -- graph.
-{-# INLINE someChainId #-}
-
--- | Uniformily get a random ChainId
---
-randomChainId :: HasChainwebVersion v => v -> IO ChainId
-randomChainId v = (!!) (toList cs) <$> randomRIO (0, length cs - 1)
-  where
-    cs = chainIds v
-
 -- -------------------------------------------------------------------------- --
 -- Properties of Chainweb Versions
 -- -------------------------------------------------------------------------- --
 
 -- -------------------------------------------------------------------------- --
 -- Graph
 
-chainwebVersionGraph :: ChainwebVersion -> ChainGraph
-chainwebVersionGraph (Test g) = g
-chainwebVersionGraph (TimedConsensus g) = g
-chainwebVersionGraph (PowConsensus g) = g
-chainwebVersionGraph (TimedCPM g) = g
-chainwebVersionGraph (FastTimedCPM g) = g
-chainwebVersionGraph Development = petersonChainGraph
-chainwebVersionGraph Testnet04 = petersonChainGraph
-chainwebVersionGraph Mainnet01 = petersonChainGraph
-
-instance HasChainGraph ChainwebVersion where
-    _chainGraph = chainwebVersionGraph
+-- | Graphs of chainweb version
+--
+-- Invariants:
+--
+-- * Entries are sorted by 'BlockHeight' in decreasing order.
+-- * The last entry is for 'BlockHeight' 0.
+-- * The graphs decrease in order.
+--
+-- The functions provided in 'Chainweb.Version.Utils' are generally more
+-- convenient to use than this function.
+--
+chainwebGraphs :: ChainwebVersion -> NE.NonEmpty (BlockHeight, ChainGraph)
```

Comment:
I'm not understanding why we have `NonEmpty (BlockHeight, ChainGraph)` here and `Map BlockHeight ChainGraph` in Utils. Both approaches lean on invariants to validate unsafe operations: here the usages of `NE.dropWhile`, in Utils `fromJuste . lookup`. 

Wouldn't it be better to have one so that the codebases of this and Utils would overlap more? `Map` seems like the better fit IMO

- the small size (and unsafety acceptance) justifies using `lookupMin`, vs `NE.head` here. 
- `fromJuste . lookupLE` seems far better than `NE.dropWhile ((> h) . fst)`

Reply:
I am not so concerned about the invariants here. Those invariants can be tested very well, and there are tests for each of them. Also, I'd rather like the this code to crash hard with an error (most likely during testing or immediately when a node is started) than somehow continuing to run if something goes wrong with these definitions. 

I think, it's an example where partial functions (with `error`) are actually making the code safer. It's the closest we get to enforcing the invariants statically, without using dependent types.

Reply:
I am using `List.NonEmpty` here because, I expect that it will give very efficient access to the head of the list, which will probably be 99.99% of all use cases. It also supports a concise human readable encoding in the code without using tricks like Template Haskell. I think, both is important for this module, which I consider the heart of chainweb.

I consider the functions in `Chainweb.Version.Utils` less mission critical. The use of `Map` was based on convenience and concise code, and the shortest path to a working implementation. Otherwise, it's total overkill and probably inefficient. There is a chance that we have replace the use of `Map` if we find that it is too inefficient, which would be fine.

Reply:
As discussed offline, this will most likely be replaced by plain `[]` in a future PR that is going to address the relation of `Chainweb.Version` and `Chainweb.Version.Utils` in more detail.